### PR TITLE
DVPN-111 - Add view of stripe report with state information

### PR DESCRIPTION
**Pull request by relud**
[DVPN-111](https://mozilla-hub.atlassian.net/browse/DVPN-111)

### CODE REVIEW

File: sql/moz-fx-data-shared-prod/stripe/itemized_payout_reconciliation/view.sql [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    STRUCT("PR" AS card_country, NULL AS state)
+  -- virgin islands
+  WHEN
+    card_country = "US"
+    AND REGEXP_CONTAINS(postal_code, "^008[0-9]{2}(-?[0-9]{4})?$")
+  THEN
+    STRUCT("VI" AS card_country, NULL AS state)
+  ELSE
+    STRUCT(card_country, state)
+  END
```
Comment (relud) [inline sql/moz-fx-data-shared-prod/stripe/itemized_payout_reconciliation/view.sql:84]: see https://mozilla-hub.atlassian.net/browse/DVPN-106?focusedCommentId=515873 for why this is needed

File: sql/moz-fx-data-shared-prod/stripe/itemized_payout_reconciliation/view.sql [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    payment_method_details.card.country = postal_code_to_state.country
+    AND billing_details.address.postal_code = postal_code_to_state.postal_code
+  WHERE
+    payment_method_details.card.country IN ("US", "CA")
+),
+enriched AS (
+  SELECT
+    report.*,
+    charges.state,
+    charges.postal_code,
```
Comment (relud) [inline sql/moz-fx-data-shared-prod/stripe/itemized_payout_reconciliation/view.sql:52]: see https://mozilla-hub.atlassian.net/browse/DVPN-111?focusedCommentId=517297 for discussion on how these fields are being calculated

**Review by sean-rose — APPROVED** [review]

File: sql/moz-fx-data-shared-prod/stripe/itemized_payout_reconciliation/view.sql (hunk 1 above)
Comment (relud) [reply sql/moz-fx-data-shared-prod/stripe/itemized_payout_reconciliation/view.sql:83]: indeed, postal code needs to be trimmed

File: sql/moz-fx-data-shared-prod/stripe/itemized_payout_reconciliation/view.sql [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    postal_code IS NOT NULL
+  GROUP BY
+    country,
+    postal_code
+),
+charges AS (
+  SELECT
+    id AS charge_id,
+    payment_method_details.card.country AS card_country,
+    NULLIF(billing_details.address.postal_code, "") AS postal_code,
```
[the remark this answers is not in the document]
Comment (relud) [reply sql/moz-fx-data-shared-prod/stripe/itemized_payout_reconciliation/view.sql:36]: `billing_details.address.country` is not available, and `billing_details.address.postal_code` is expected to be within `payment_method_details.card.country` where present

File: sql/moz-fx-data-shared-prod/stripe/itemized_payout_reconciliation/view.sql [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  `moz-fx-data-shared-prod.stripe.itemized_payout_reconciliation`
+AS
+WITH customers AS (
+  SELECT
+    id,
+    address.country,
+    NULLIF(address.postal_code, "") AS postal_code,
+    NULLIF(address.state, "") AS state,
+  FROM
+    `moz-fx-data-shared-prod`.stripe_external.customers_v1
```
[the remark this answers is not in the document]
Comment (relud) [reply sql/moz-fx-data-shared-prod/stripe/itemized_payout_reconciliation/view.sql:11]: in DVPN-111 i pointed out that this is not preferable long-term, and should be replaced with a proper lookup table.

File: sql/moz-fx-data-shared-prod/stripe/itemized_payout_reconciliation/view.sql [hunk 5]
```diff
@@ -0,0 +1,86 @@
+CREATE OR REPLACE VIEW
+  `moz-fx-data-shared-prod.stripe.itemized_payout_reconciliation`
+AS
+WITH customers AS (
```
[the remark this answers is not in the document]
Comment (relud) [reply sql/moz-fx-data-shared-prod/stripe/itemized_payout_reconciliation/view.sql:4]: will rename

File: sql/moz-fx-data-shared-prod/stripe/itemized_payout_reconciliation/view.sql [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+),
+postal_code_to_state AS (
+  SELECT
+    country,
+    postal_code,
+    IF(
+      COUNT(DISTINCT state) > 1,
+      ERROR("postal code resolved to mulitple states: " || TO_JSON_STRING(postal_code)),
+      ANY_VALUE(state)
+    ) AS state,
```
[the remark this answers is not in the document]
Comment (relud) [reply sql/moz-fx-data-shared-prod/stripe/itemized_payout_reconciliation/view.sql:23]: fair enough. I know that when subplat eng gets multiple possible states for a postal code, they used NULL, so i'll follow that approach.

**Pull request merged**